Implement shopping cart functionality and ticket persistence

flashforward55 opened the pull request.

flashforward55 committed:
Implement shopping cart functionality and ticket persistence

mateacademy-ai-mentor commented on src/main/java/mate/academy/dao/impl/ShoppingCartDaoImpl.java:
```diff
@@ -0,0 +1,80 @@
+            transaction = session.beginTransaction();
+            User managedUser = session.merge(shoppingCart.getUser());
+            shoppingCart.setUser(managedUser);
+            session.persist(shoppingCart);
```
This violates checklist item #5: 'For saving ShoppingCart use `save()` method instead of `persist()`.' The `persist()` method can throw an exception if the associated `User` entity is in a detached state. Using `save()` handles this case correctly.

mateacademy-ai-mentor commented on src/main/java/mate/academy/dao/impl/ShoppingCartDaoImpl.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Override
+    public ShoppingCart getByUser(User user) {
```
According to the task description for the `ShoppingCartDao` interface, this method should return `Optional<ShoppingCart>`. Please update the method signature and the implementation to wrap the result of `query.uniqueResult()` in an `Optional`.

mateacademy-ai-mentor commented on src/main/java/mate/academy/dao/ShoppingCartDao.java:
```diff
@@ -0,0 +1,12 @@
+public interface ShoppingCartDao {
+    ShoppingCart add(ShoppingCart shoppingCart);
+
+    ShoppingCart getByUser(User user);
```
The return type of this method should be `Optional<ShoppingCart>` as specified in the task description. Using `Optional` is a good practice to explicitly handle cases where a shopping cart for a given user might not be found, thus preventing potential `NullPointerException`s.

mateacademy-ai-mentor commented on src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+
+    @Override
+    public ShoppingCart getByUser(User user) {
+        return shoppingCartDao.getByUser(user);
```
According to the task description, `ShoppingCartDao.getByUser()` should return an `Optional<ShoppingCart>`, while this service method should return a `ShoppingCart`. This means you need to handle the `Optional` here. You should get the value from the optional or throw an exception if it's not present.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

This requires a couple of changes to meet the task requirements. Your overall logic is solid, but there are two specific issues in the DAO layer that need to be addressed.

1.  In `ShoppingCartDao.java` and `ShoppingCartDaoImpl.java`, the `getByUser()` method must return `Optional<ShoppingCart>`. This is a key requirement to handle cases where a cart might not exist. You'll also need to update `ShoppingCartServiceImpl` to correctly handle the `Optional` returned from the DAO.
2.  In `ShoppingCartDaoImpl.java`, the `add()` method should use `session.save()` instead of `session.persist()`. This is required by the checklist [CHECKLIST ITEM #5] to prevent potential exceptions when the associated `User` entity is in a detached state.

Once you've made these adjustments, your solution will be ready. Great work on the rest of the implementation!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

flashforward55 committed:
fix: use Optional for getByUser and save() instead of persist()

- Change ShoppingCartDao.getByUser() to return Optional<ShoppingCart>
- Replace session.persist() with session.save() in add() method
- Handle Optional in ShoppingCartServiceImpl with orElseThrow()
- Prevent NullPointerException and detached entity issues

mateacademy-ai-mentor commented on src/main/java/mate/academy/dao/impl/TicketDaoImpl.java:
```diff
@@ -0,0 +1,34 @@
+        try {
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.persist(ticket);
```
The `persist()` method is likely to throw an exception here. The `ticket` object holds references to `User` and `MovieSession` objects, which are probably in a 'detached' state. The `persist()` method doesn't work with detached entities. The same reasoning from the checklist about using `save()` for `ShoppingCart` applies here. Consider using a method that can handle this, such as `save()` or `merge()`.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Great progress on this task! You've successfully addressed the previous feedback, and most of the implementation is excellent. However, there is one critical issue in your `TicketDaoImpl` that needs to be fixed before this can be approved.

In the `add` method of `src/main/java/mate/academy/dao/impl/TicketDaoImpl.java`, you are using `session.persist(ticket)`. This will cause a runtime exception because the `ticket` object contains `User` and `MovieSession` entities that are in a detached state. To resolve this, please use a method that correctly handles saving entities with detached associations, such as `session.save()`. You've already implemented the correct approach for the `ShoppingCart`, so you can apply the same logic here.

Once you make this change, the solution will be ready for approval. Keep up the great work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

flashforward55 committed:
fix: use save() instead of persist() in TicketDaoImpl

- Replace session.persist() with session.save() in add() method
- Prevent exceptions when Ticket contains detached User and MovieSession
- Apply same approach as in ShoppingCartDaoImpl

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Excellent work on this task! I am approving your solution as it perfectly meets all the requirements. You've successfully implemented the `ShoppingCart` and `Ticket` models with the correct entity relationships using `@OneToOne`, `@OneToMany`, and `@MapsId`. Your DAO and service layers are well-structured, demonstrating a solid understanding of Hibernate, transaction management, and proper resource handling. The integration into the authentication flow is also flawless. This is a high-quality submission, great job!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
